### DISCUSSION TITLE
How to get these matrices?

### DISCUSSION THREAD

**Discussion by icnosvbovwb** [OpenFAST/openfast#1513] [Q&A]
Dear @jjonkman 
I want to obtain the matrix of these four types of wind turbines, Can you give me a specific piece of information？
 ITI Barge,OC3spar, TLP, OC4 SEMI
 
 Body Mass Matrix:
 External Linear Damping Matrix:
 External Quadratic Damping Matrix:
 Hydrostatic Restoring Matrix:
 External Restoring Matrix:

**Comment by jjonkman** [comment]
Dear @icnosvbovwb,

Here are a few responses:

- Body Mass Matrix -- See the following topic on our forum for information on how to use the linearization funcationality of OpenFAST to extract the 6x6 rigid-body mass matrix of ElastoDyn: https://forums.nrel.gov/t/openfast-2nd-order-linearization/2249.
- External Linear Damping Matrix -- If you mean that you want to know what is specified as `AddBLin` in HydroDyn, see the corresponding HydroDyn input file for each model (available in the OpenFAST r-test: https://github.com/OpenFAST/r-test).  If you want to know the equivalent linear damping from all contributions (hydro, aero), you'd have to use the linearization functionality of OpenFAST.
- External Quadratic Damping Matrix -- If you mean that you want to know what is specified as `AddBQuad` in HydroDyn, see the corresponding HydroDyn input file for each model.  If you want to know how to represent distributed viscous drag from the strip-theory solution of HydroDyn as a 6x6 matrix or similar for aerodynamics, I've not done that myself.
- Hydrostatic Restoring Matrix -- If you mean that you want to know what is specified as the linear hydrostatic restoring matrix in the potential-flow solution in HydroDyn, see the corresponding WAMIT _.hst_ file for each model (note that this data is nondimensional and must be dimensionalized if you want that).  Of course, other terms can effect the total hydrostatic restoring, including the impact of platform weight and ballast that are accounted for elsewhere within OpenFAST.  The total 6x6 stiffness matrix can be derived through OpenFAST linearization.
- External Restoring Matrix -- If you mean the mooring system restoring, this can be obtained, e.g., through MAP++ as discussed on our forum: https://forums.nrel.gov/t/mooring-line-stiffness-matrix/3785.  Again, OpenFAST linearization can derived the full-system 6x6 stiffness matrix including all contributions.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/229768328-3bd281f7-4c81-4e5f-a9cc-1d66d0d241fb.png)
![image](https://user-images.githubusercontent.com/107483730/229769332-42c0a8ae-8852-4013-9b94-c4fe8952702d.png)
![image](https://user-images.githubusercontent.com/107483730/229769286-081c9175-2895-4bae-a9b9-2d1ba8dfd7e5.png)
I extract the six numbers of the D matrix in the data, perform diagonalization, and then perform inversion. The resulting 6x6 matrix is different. What other operations do I need to perform?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

You should not have to perform "diagonalization" of the D matrix.  I suspect that you are not extracting the correct 6x6 submatrix from D.  To find the 6x6 mass matrix of ElastoDyn, you must extract the 6x6 submatrix of D associated with the 6 platform load inputs and 6 platform acceleration outputs.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Q1：May I ask what is the difference between linearization and nonlinearity? Although there are differences in my calculations, the difference is not significant, or what are the advantages of linearization?
![image](https://github.com/OpenFAST/openfast/assets/107483730/5972dd7b-8789-4d47-8bbd-ce1bbec77339)
![image](https://github.com/OpenFAST/openfast/assets/107483730/f9195ee3-7fd6-4212-8f7f-d27dfa1f56b7)

Q2：Can you tell me specifically how to obtain these matrices through linearization?
Body Mass Matrix:
External Linear Damping Matrix:
External Quadratic Damping Matrix:
Hydrostatic Restoring Matrix:
External Restoring Matrix:
Because I can easily manipulate the linearization process, but I am not sure how to implement these Matrix.

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

For a nonlinear model like OpenFAST, a linear model is only valid for small perturbations about the operating point.  The advantage of a linear model is that you can really understand the system response and do things such as modal analysis.

Regarding the matrices you mention, are you referring to 6x6 matrices associated with rigid-body motion of floating platform?  Quadratic drag is not linear, so, the linearization process can't derive that.  The full-system linearization of OpenFAST can derive the full-system linearized mass, damping, and stiffness/restoring matrices, but it is difficult to separate out the contributions (between hydrostatic, mooring, body weight, etc.) unless (1) you already know what some terms are so that you can derive the remaining terms by subtracting the known terms from the total or (2) you use the module-level linearized matrices to isolate the effects from each module.  The following topic on our forum explains how to derive the full-system 6x6 mass, stiffness, and damping matrices through OpenFAST linearization: https://forums.nrel.gov/t/openfast-2nd-order-linearization/2249.

Best regards,
